src/services/urlAssembly.ts

```diff
@@ -19,23 +22,25 @@ type AssembleUrlOptions = {
 export function assembleUrl(route: Route, options: AssembleUrlOptions = {}): Url {
   const { params: paramValues = {}, query: queryValues } = options
 
-  const hostWithParamsSet = assembleHostParamValues(route.host, paramValues)
   const pathWithParamsSet = assemblePathParamValues(route.path, paramValues)
   const queryWithParamsSet = assembleQueryParamValues(route.query, paramValues)
+  const hostWithParamsSet = assembleHostParamValues(route.host, paramValues)
+  const hash = createHash(route.hash.value ?? options.hash)
 
-  const hostPathAndQuery = withQuery(`${hostWithParamsSet}${pathWithParamsSet}`, queryWithParamsSet, queryValues)
-  const hash = createHash(route.hash.value ?? options.hash).toString()
-  const url = `${hostPathAndQuery}${hash}`
+  type UrlAssemblyFunction = (url: Url) => Url
 
-  if (isUrl(url)) {
-    return url
-  }
+  const assembly: UrlAssemblyFunction[] = [
+    (url) => withPath(url, pathWithParamsSet),
+    (url) => withQuery(url, queryWithParamsSet, queryValues),
+    (url) => withHost(url, hostWithParamsSet),
+    (url) => withHash(url, hash),
+  ]
 
-  return `/${url}`
+  return assembly.reduce<Url>((url, join) => join(url), '/')
 }
 
 function assembleHostParamValues(host: Host, paramValues: Record<string, unknown>): string {
-  const value = host.toString()
+  const { value } = host
```

Comment:
nit: should probably just use `host.value` rather than destructuring. Same for path and query.